// @celo/identity/lib/offchain/accessors/authorized-signer.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.AuthorizedSignerAccessor = void 0;
const base_1 = require("@celo/base");
const result_1 = require("@celo/base/lib/result");
const address_1 = require("@celo/utils/lib/address");
const io_1 = require("@celo/utils/lib/io");
const t = __importStar(require("io-ts"));
const utils_1 = require("../utils");
const errors_1 = require("./errors");
const AuthorizedSignerSchema = t.type({
    address: io_1.AddressType,
    proofOfPossession: io_1.SignatureType,
    filteredDataPaths: t.string,
});
class AuthorizedSignerAccessor {
    constructor(wrapper) {
        this.wrapper = wrapper;
        this.basePath = '/account/authorizedSigners';
        this.read = (0, result_1.makeAsyncThrowable)(this.readAsResult.bind(this));
    }
    readAsResult(account, signer) {
        return __awaiter(this, void 0, void 0, function* () {
            const dataPath = this.basePath + '/' + (0, address_1.toChecksumAddress)(signer);
            const rawData = yield this.wrapper.readDataFromAsResult(account, dataPath, false, AuthorizedSignerSchema);
            if (!rawData.ok) {
                return (0, result_1.Err)(new errors_1.OffchainError(rawData.error));
            }
            return (0, utils_1.deserialize)(AuthorizedSignerSchema, rawData.result);
        });
    }
    write(signer, proofOfPossession, filteredDataPaths) {
        return __awaiter(this, void 0, void 0, function* () {
            const payload = {
                address: (0, address_1.toChecksumAddress)(signer),
                proofOfPossession,
                filteredDataPaths,
            };
            const dataPath = this.basePath + '/' + (0, address_1.toChecksumAddress)(signer);
            const typedData = yield (0, utils_1.buildEIP712TypedData)(this.wrapper, dataPath, payload, AuthorizedSignerSchema);
            const signature = yield this.wrapper.kit
                .getWallet()
                .signTypedData(this.wrapper.self, typedData);
            return this.wrapper.writeDataTo(Buffer.from(JSON.stringify(payload)), Buffer.from((0, base_1.trimLeading0x)(signature), 'hex'), dataPath);
        });
    }
}
exports.AuthorizedSignerAccessor = AuthorizedSignerAccessor;
