// @celo/identity/lib/offchain/accessors/binary.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.PrivateBinaryAccessor = exports.PublicBinaryAccessor = void 0;
const address_1 = require("@celo/base/lib/address");
const result_1 = require("@celo/base/lib/result");
const utils_1 = require("../utils");
const errors_1 = require("./errors");
/**
 * Schema for writing any generic binary data
 */
class PublicBinaryAccessor {
    constructor(wrapper, dataPath) {
        this.wrapper = wrapper;
        this.dataPath = dataPath;
        this.read = (0, result_1.makeAsyncThrowable)(this.readAsResult.bind(this));
    }
    write(data) {
        return __awaiter(this, void 0, void 0, function* () {
            const signature = yield (0, utils_1.signBuffer)(this.wrapper, this.dataPath, data);
            const error = yield this.wrapper.writeDataTo(data, Buffer.from((0, address_1.trimLeading0x)(signature), 'hex'), this.dataPath);
            if (error) {
                return new errors_1.OffchainError(error);
            }
        });
    }
    readAsResult(account) {
        return __awaiter(this, void 0, void 0, function* () {
            const rawData = yield this.wrapper.readDataFromAsResult(account, this.dataPath, true);
            if (!rawData.ok) {
                return (0, result_1.Err)(new errors_1.OffchainError(rawData.error));
            }
            return (0, result_1.Ok)(rawData.result);
        });
    }
}
exports.PublicBinaryAccessor = PublicBinaryAccessor;
/**
 * Schema for writing any encrypted binary data.
 */
class PrivateBinaryAccessor {
    constructor(wrapper, dataPath) {
        this.wrapper = wrapper;
        this.dataPath = dataPath;
        this.read = (0, result_1.makeAsyncThrowable)(this.readAsResult.bind(this));
    }
    write(data, toAddresses, symmetricKey) {
        return __awaiter(this, void 0, void 0, function* () {
            return (0, utils_1.writeEncrypted)(this.wrapper, this.dataPath, data, toAddresses, symmetricKey);
        });
    }
    allowAccess(toAddresses, symmetricKey) {
        return __awaiter(this, void 0, void 0, function* () {
            return (0, utils_1.writeSymmetricKeys)(this.wrapper, this.dataPath, toAddresses, symmetricKey);
        });
    }
    readAsResult(account) {
        return __awaiter(this, void 0, void 0, function* () {
            return (0, utils_1.readEncrypted)(this.wrapper, this.dataPath, account);
        });
    }
}
exports.PrivateBinaryAccessor = PrivateBinaryAccessor;
