// @celo/identity/lib/odis/bls-blinding-client.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.WasmBlsBlindingClient = void 0;
const crypto_1 = require("crypto");
class WasmBlsBlindingClient {
    constructor(odisPubKey) {
        // https://stackoverflow.com/questions/17575790/environment-detection-node-js-or-browser
        this.isBrowserEnvironment = new Function('try {return this===window;}catch(e){ return false;}');
        this.odisPubKey = Buffer.from(odisPubKey, 'base64');
        // Dynamically load the Wasm library
        // Checkout out documentation for alternative runtime environments:
        // https://github.com/celo-org/identity/tree/ASv2/asv2#runtime-environments
        if (this.isReactNativeEnvironment()) {
            throw new Error('Cannot use WasmBlsBlindingClient in a React Native app');
        }
        else if (this.isBrowserEnvironment()) {
            throw new Error('Cannot use WasmBlsBlindingClient in a browser environment');
        }
        else {
            this.thresholdBls = require('blind-threshold-bls');
        }
    }
    blindMessage(base64PhoneNumber, seed) {
        return __awaiter(this, void 0, void 0, function* () {
            const userSeed = seed !== null && seed !== void 0 ? seed : (0, crypto_1.randomBytes)(32);
            if (!seed) {
                console.warn('Warning: Use a private deterministic seed (e.g. DEK private key) to preserve user quota when requests are replayed.');
            }
            this.rawMessage = Buffer.from(base64PhoneNumber, 'base64');
            this.blindedValue = yield this.thresholdBls.blind(this.rawMessage, userSeed);
            const blindedMessage = this.blindedValue.message;
            return Buffer.from(blindedMessage).toString('base64');
        });
    }
    unblindAndVerifyMessage(base64BlindSig) {
        return __awaiter(this, void 0, void 0, function* () {
            if (!this.rawMessage || !this.blindedValue) {
                throw new Error('Must call blind before unblinding');
            }
            const blindedSignature = Buffer.from(base64BlindSig, 'base64');
            const unblindMessage = yield this.thresholdBls.unblind(blindedSignature, this.blindedValue.blindingFactor);
            // this throws on error
            yield this.thresholdBls.verify(this.odisPubKey, this.rawMessage, unblindMessage);
            return Buffer.from(unblindMessage).toString('base64');
        });
    }
    isReactNativeEnvironment() {
        return typeof navigator !== 'undefined' && navigator.product === 'ReactNative';
    }
}
exports.WasmBlsBlindingClient = WasmBlsBlindingClient;
