// @celo/identity/lib/odis/circuit-breaker.mock.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.MockCircuitBreaker = exports.MOCK_CIRCUIT_BREAKER_PRIVATE_KEY = exports.MOCK_CIRCUIT_BREAKER_PUBLIC_KEY = void 0;
const crypto = __importStar(require("crypto"));
const debug_1 = __importDefault(require("debug"));
const circuit_breaker_1 = require("./circuit-breaker");
const debug = (0, debug_1.default)('kit:identity:odis:circuit-breaker:mock');
exports.MOCK_CIRCUIT_BREAKER_PUBLIC_KEY = `-----BEGIN PUBLIC KEY-----
MIGfMA0GCSqGSIb3DQEBAQUAA4GNADCBiQKBgQDGtxUPljt+oHFBf3RrDZHN9TbT
iI0kK4bv02Z2WP7kU/PQCikWqNl9/VjGVXuGMlfwpcWZrjWwJa+kBlUYRXH/inXW
UKO5PqTnaUXS1ALasGAUvzRz3VvzCxpjKsjVS8/gAoJbY2Imwor432OLrOssNoK7
jbl1TgaV47yGCKwF9wIDAQAB
-----END PUBLIC KEY-----`;
exports.MOCK_CIRCUIT_BREAKER_PRIVATE_KEY = `-----BEGIN PRIVATE KEY-----
MIICdwIBADANBgkqhkiG9w0BAQEFAASCAmEwggJdAgEAAoGBAMa3FQ+WO36gcUF/
dGsNkc31NtOIjSQrhu/TZnZY/uRT89AKKRao2X39WMZVe4YyV/ClxZmuNbAlr6QG
VRhFcf+KddZQo7k+pOdpRdLUAtqwYBS/NHPdW/MLGmMqyNVLz+ACgltjYibCivjf
Y4us6yw2gruNuXVOBpXjvIYIrAX3AgMBAAECgYBGPqv8QZAweAjxLVv7B+X112wV
JN033wcpOiKrTVR1ZFP4w864iuGvTuKV4dvzmVJK6F7Mr6+c4AWRxwdHuCzOlwxj
O9RySFAXhoENu70zg8W2w4i8GMHsmdnNk045cF01Mb3GtQ6Y3uGb637XYTIwMEbC
Q74TbkrfPZPcSIpPEQJBAP4VModTr47oNvdyJITQ3fzIarRSDU0deZTpn6MXB3a1
abOAzlqYK3CSvLyyM9GOB9C5wvIZev+aNU9SkqPzU38CQQDINu7nOqS2X8UXQ5sS
wFrnoBQcU78i7Jaopvw0kOvkvklHlKVvXVkWP8PaWYdUAO9fpEdKdRnfaOEnqBwT
aymJAkEAgTXmbEtyjAoracryJ1jQiyyglvLjMMQ8gC4OsLGVahj3mAF47zlTXfxB
XvSAxaCk+NB/Av9SPYn+ckhbqmSjoQJAYb6H1bVIkoyg0OG9hGMKPkhlaQrtpmQw
jTewqw0RTQQlDGAigALnqjgJKsFIkxc9xciS0WPn9KzkNxMYWdaYWQJBAI8asXXb
XF5Lg2AAM2xJ/SS+h+si4f70eZey4vo9pWB3Q+VKbtRZu2pCjlR1A1nIqigJxdlc
1jHX+4GiW+t0w8Q=
-----END PRIVATE KEY-----`;
const MOCK_CIRCUIT_BREAKER_ENVIRONMENT = {
    url: 'https://mockcircuitbreaker.com/',
    publicKey: exports.MOCK_CIRCUIT_BREAKER_PUBLIC_KEY,
};
/**
 * Mock circuit breaker implementation based on Valora implementaion
 * github.com/valora-inc/wallet/tree/main/packages/cloud-functions/src/circuitBreaker/circuitBreaker.ts
 */
class MockCircuitBreaker {
    constructor() {
        this.keyStatus = circuit_breaker_1.CircuitBreakerKeyStatus.ENABLED;
    }
    status() {
        return {
            status: 200,
            body: { status: this.keyStatus },
        };
    }
    unwrapKey(req) {
        const { ciphertext } = req;
        if (!ciphertext) {
            return {
                status: 400,
                body: { error: '"ciphertext" parameter must be provided' },
            };
        }
        else if (!circuit_breaker_1.BASE64_REGEXP.test(ciphertext)) {
            return {
                status: 400,
                body: { error: '"ciphertext" parameter must be a base64 encoded buffer' },
            };
        }
        if (this.keyStatus !== circuit_breaker_1.CircuitBreakerKeyStatus.ENABLED) {
            return {
                status: 503,
                body: { status: this.keyStatus },
            };
        }
        let plaintext;
        try {
            plaintext = crypto.privateDecrypt(
            // @ts-ignore support for OAEP hash option, was added in Node 12.9.0.
            { key: exports.MOCK_CIRCUIT_BREAKER_PRIVATE_KEY, oaepHash: 'sha256' }, Buffer.from(ciphertext, 'base64'));
        }
        catch (error) {
            return {
                status: 500,
                body: { error: 'Error while decrypting ciphertext' },
            };
        }
        return {
            status: 200,
            body: { plaintext: plaintext.toString('base64') },
        };
    }
    installStatusEndpoint(mock, override) {
        mock.mock({
            url: new URL(circuit_breaker_1.CircuitBreakerEndpoints.STATUS, MockCircuitBreaker.environment.url).href,
            method: 'GET',
        }, override !== null && override !== void 0 ? override : ((url, req) => {
            debug('Mocking request', { url, req });
            return this.status();
        }));
    }
    installUnwrapKeyEndpoint(mock, override) {
        mock.mock({
            url: new URL(circuit_breaker_1.CircuitBreakerEndpoints.UNWRAP_KEY, MockCircuitBreaker.environment.url).href,
            method: 'POST',
        }, override !== null && override !== void 0 ? override : ((url, req) => {
            debug('Mocking request', { url, req });
            return this.unwrapKey(JSON.parse(req.body));
        }));
    }
    install(mock) {
        this.installStatusEndpoint(mock);
        this.installUnwrapKeyEndpoint(mock);
    }
}
exports.MockCircuitBreaker = MockCircuitBreaker;
MockCircuitBreaker.publicKey = exports.MOCK_CIRCUIT_BREAKER_PUBLIC_KEY;
MockCircuitBreaker.privateKey = exports.MOCK_CIRCUIT_BREAKER_PRIVATE_KEY;
MockCircuitBreaker.environment = MOCK_CIRCUIT_BREAKER_ENVIRONMENT;
