// @celo/identity/lib/odis/phone-number-identifier.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.isBalanceSufficientForSigRetrieval = exports.getPhoneNumberIdentifierFromSignature = exports.getBlindedPhoneNumberSignature = exports.getBlindedPhoneNumber = exports.getPhoneNumberIdentifier = exports.ODIS_MINIMUM_CELO_BALANCE = exports.ODIS_MINIMUM_DOLLAR_BALANCE = void 0;
const bignumber_js_1 = __importDefault(require("bignumber.js"));
const debug_1 = __importDefault(require("debug"));
const identifier_1 = require("./identifier");
// ODIS minimum dollar balance for sig retrieval
exports.ODIS_MINIMUM_DOLLAR_BALANCE = 0.01;
// ODIS minimum celo balance for sig retrieval
exports.ODIS_MINIMUM_CELO_BALANCE = 0.005;
const debug = (0, debug_1.default)('kit:odis:phone-number-identifier');
/**
 * Retrieve the on-chain identifier for the provided phone number
 * Performs blinding, querying, and unblinding
 * @deprecated use getObfuscatedIdentifier instead
 */
function getPhoneNumberIdentifier(e164Number, account, signer, context, blindingFactor, clientVersion, blsBlindingClient, sessionID, keyVersion) {
    return __awaiter(this, void 0, void 0, function* () {
        debug('Getting phone number pepper');
        const { plaintextIdentifier, obfuscatedIdentifier, pepper, unblindedSignature } = yield (0, identifier_1.getObfuscatedIdentifier)(e164Number, identifier_1.IdentifierPrefix.PHONE_NUMBER, account, signer, context, blindingFactor, clientVersion, blsBlindingClient, sessionID, keyVersion);
        return {
            e164Number: plaintextIdentifier,
            phoneHash: obfuscatedIdentifier,
            pepper,
            unblindedSignature,
        };
    });
}
exports.getPhoneNumberIdentifier = getPhoneNumberIdentifier;
/**
 * Blinds the phone number in preparation for the ODIS request
 * Caller should use the same blsBlindingClient instance for unblinding
 * @deprecated use getBlindedIdentifier instead
 */
function getBlindedPhoneNumber(e164Number, blsBlindingClient, seed) {
    return __awaiter(this, void 0, void 0, function* () {
        return (0, identifier_1.getBlindedIdentifier)(e164Number, identifier_1.IdentifierPrefix.PHONE_NUMBER, blsBlindingClient, seed);
    });
}
exports.getBlindedPhoneNumber = getBlindedPhoneNumber;
/**
 * Query ODIS for the blinded signature
 * Response can be passed into getPhoneNumberIdentifierFromSignature
 * to retrieve the on-chain identifier
 * @deprecated use getBlindedIdentifierSignature instead
 */
function getBlindedPhoneNumberSignature(account, signer, context, base64BlindedMessage, clientVersion, sessionID, keyVersion) {
    return __awaiter(this, void 0, void 0, function* () {
        return (0, identifier_1.getBlindedIdentifierSignature)(account, signer, context, base64BlindedMessage, clientVersion, sessionID, keyVersion);
    });
}
exports.getBlindedPhoneNumberSignature = getBlindedPhoneNumberSignature;
/**
 * Unblind the response and return the on-chain identifier
 * @deprecated use getObfuscatedIdentifieriFromSignature instead
 */
function getPhoneNumberIdentifierFromSignature(e164Number, base64BlindedSignature, blsBlindingClient) {
    return __awaiter(this, void 0, void 0, function* () {
        const { plaintextIdentifier, obfuscatedIdentifier, pepper, unblindedSignature } = yield (0, identifier_1.getObfuscatedIdentifierFromSignature)(e164Number, identifier_1.IdentifierPrefix.PHONE_NUMBER, base64BlindedSignature, blsBlindingClient);
        return {
            e164Number: plaintextIdentifier,
            phoneHash: obfuscatedIdentifier,
            pepper,
            unblindedSignature,
        };
    });
}
exports.getPhoneNumberIdentifierFromSignature = getPhoneNumberIdentifierFromSignature;
/**
 * Check if balance is sufficient for quota retrieval
 * @deprecated use getPnpQuotaStatus instead
 */
function isBalanceSufficientForSigRetrieval(dollarBalance, celoBalance) {
    return (new bignumber_js_1.default(dollarBalance).isGreaterThanOrEqualTo(exports.ODIS_MINIMUM_DOLLAR_BALANCE) ||
        new bignumber_js_1.default(celoBalance).isGreaterThanOrEqualTo(exports.ODIS_MINIMUM_CELO_BALANCE));
}
exports.isBalanceSufficientForSigRetrieval = isBalanceSufficientForSigRetrieval;
