// @celo/identity/lib/odis/quota.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getPnpQuotaStatus = void 0;
const phone_number_privacy_common_1 = require("@celo/phone-number-privacy-common");
const query_1 = require("./query");
/**
 * Query the ODIS quota status of a given account
 *
 * @param account The address whose ODIS quota we are querying
 * @param signer Object containing the private key used to authenticate the ODIS request
 * @param context Specifies which ODIS combiner url should be queried (i.e. mainnet or alfajores)
 * @param clientVersion Optional Specifies the client software version
 * @param sessionID Optional Used to track user sessions across the client and ODIS
 * @param abortController Optional Allows client to specify a timeout for the ODIS request
 */
function getPnpQuotaStatus(account, signer, context, clientVersion, sessionID, abortController) {
    return __awaiter(this, void 0, void 0, function* () {
        const body = {
            account,
            version: clientVersion,
            authenticationMethod: signer.authenticationMethod,
            sessionID,
        };
        const response = yield (0, query_1.queryOdis)(body, context, phone_number_privacy_common_1.CombinerEndpoint.PNP_QUOTA, phone_number_privacy_common_1.PnpQuotaResponseSchema, {
            Authorization: yield (0, query_1.getOdisPnpRequestAuth)(body, signer),
        }, abortController);
        if (response.success) {
            return {
                version: response.version,
                performedQueryCount: response.performedQueryCount,
                totalQuota: response.totalQuota,
                remainingQuota: response.totalQuota - response.performedQueryCount,
                warnings: response.warnings,
            };
        }
        throw new Error(response.error);
    });
}
exports.getPnpQuotaStatus = getPnpQuotaStatus;
