// @celo/identity/lib/offchain/storage-writers.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.MockStorageWriter = exports.AwsStorageWriter = exports.GoogleStorageWriter = exports.GitStorageWriter = exports.LocalStorageWriter = exports.StorageWriter = void 0;
const child_process_1 = require("child_process");
const fs_1 = require("fs");
const path_1 = require("path");
const utils_1 = require("./utils");
class StorageWriter {
}
exports.StorageWriter = StorageWriter;
class LocalStorageWriter extends StorageWriter {
    constructor(root) {
        super();
        this.root = root;
    }
    write(data, dataPath) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.writeToFs(data, dataPath);
        });
    }
    writeToFs(data, dataPath) {
        return __awaiter(this, void 0, void 0, function* () {
            const directory = (0, path_1.parse)(dataPath).dir;
            yield fs_1.promises.mkdir((0, path_1.join)(this.root, directory), { recursive: true });
            yield fs_1.promises.writeFile((0, path_1.join)(this.root, dataPath), data);
        });
    }
}
exports.LocalStorageWriter = LocalStorageWriter;
class GitStorageWriter extends LocalStorageWriter {
    write(data, dataPath) {
        return __awaiter(this, void 0, void 0, function* () {
            yield this.writeToFs(data, dataPath);
            (0, child_process_1.spawnSync)('git', ['add', dataPath], {
                cwd: this.root,
            });
            (0, child_process_1.spawnSync)('git', ['commit', '--message', `"Upload ${dataPath}"`], { cwd: this.root });
            (0, child_process_1.spawnSync)('git', ['push', 'origin', 'master'], { cwd: this.root });
            return;
        });
    }
}
exports.GitStorageWriter = GitStorageWriter;
class GoogleStorageWriter extends LocalStorageWriter {
    constructor(local, bucket) {
        super(local);
        this.local = local;
        this.bucket = bucket;
    }
    write(data, dataPath) {
        return __awaiter(this, void 0, void 0, function* () {
            yield this.writeToFs(data, dataPath);
            (0, child_process_1.spawnSync)('gsutil', ['cp', (0, path_1.join)(this.root, dataPath), `gs://${this.bucket}${dataPath}`], {
                cwd: this.root,
            });
        });
    }
}
exports.GoogleStorageWriter = GoogleStorageWriter;
class AwsStorageWriter extends LocalStorageWriter {
    constructor(local, bucket) {
        super(local);
        this.local = local;
        this.bucket = bucket;
    }
    write(data, dataPath) {
        return __awaiter(this, void 0, void 0, function* () {
            yield this.writeToFs(data, dataPath);
            (0, child_process_1.spawnSync)('aws', ['s3', 'cp', (0, path_1.join)(this.root, dataPath), `s3://${this.bucket}${dataPath}`], {
                cwd: this.root,
            });
        });
    }
}
exports.AwsStorageWriter = AwsStorageWriter;
class MockStorageWriter extends LocalStorageWriter {
    constructor(root, mockedStorageRoot, fetchMock) {
        super(root);
        this.root = root;
        this.mockedStorageRoot = mockedStorageRoot;
        this.fetchMock = fetchMock;
    }
    write(data, dataPath) {
        return __awaiter(this, void 0, void 0, function* () {
            yield this.writeToFs(data, dataPath);
            this.fetchMock.mock((0, utils_1.resolvePath)(this.mockedStorageRoot, dataPath), data, {
                sendAsJson: false,
                overwriteRoutes: true,
            });
        });
    }
}
exports.MockStorageWriter = MockStorageWriter;
