// @celo/phone-number-privacy-common/lib/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.rootLogger = exports.loggerMiddleware = exports.genSessionID = exports.getContractKitWithAgent = exports.getContractKit = exports.toNum = exports.toBool = exports.fetchEnvOrDefault = exports.fetchEnv = exports.TestUtils = exports.ThresholdPoprfServer = exports.ThresholdPoprfClient = exports.PoprfServer = exports.PoprfCombiner = exports.PoprfClient = exports.WarningMessage = exports.ErrorMessage = void 0;
__exportStar(require("./domains"), exports);
__exportStar(require("./interfaces"), exports);
var errors_1 = require("./interfaces/errors");
Object.defineProperty(exports, "ErrorMessage", { enumerable: true, get: function () { return errors_1.ErrorMessage; } });
Object.defineProperty(exports, "WarningMessage", { enumerable: true, get: function () { return errors_1.WarningMessage; } });
var poprf_1 = require("./poprf");
Object.defineProperty(exports, "PoprfClient", { enumerable: true, get: function () { return poprf_1.PoprfClient; } });
Object.defineProperty(exports, "PoprfCombiner", { enumerable: true, get: function () { return poprf_1.PoprfCombiner; } });
Object.defineProperty(exports, "PoprfServer", { enumerable: true, get: function () { return poprf_1.PoprfServer; } });
Object.defineProperty(exports, "ThresholdPoprfClient", { enumerable: true, get: function () { return poprf_1.ThresholdPoprfClient; } });
Object.defineProperty(exports, "ThresholdPoprfServer", { enumerable: true, get: function () { return poprf_1.ThresholdPoprfServer; } });
var index_1 = require("./test/index");
Object.defineProperty(exports, "TestUtils", { enumerable: true, get: function () { return index_1.TestUtils; } });
__exportStar(require("./utils/authentication"), exports);
var config_utils_1 = require("./utils/config.utils");
Object.defineProperty(exports, "fetchEnv", { enumerable: true, get: function () { return config_utils_1.fetchEnv; } });
Object.defineProperty(exports, "fetchEnvOrDefault", { enumerable: true, get: function () { return config_utils_1.fetchEnvOrDefault; } });
Object.defineProperty(exports, "toBool", { enumerable: true, get: function () { return config_utils_1.toBool; } });
Object.defineProperty(exports, "toNum", { enumerable: true, get: function () { return config_utils_1.toNum; } });
__exportStar(require("./utils/constants"), exports);
var contracts_1 = require("./utils/contracts");
Object.defineProperty(exports, "getContractKit", { enumerable: true, get: function () { return contracts_1.getContractKit; } });
Object.defineProperty(exports, "getContractKitWithAgent", { enumerable: true, get: function () { return contracts_1.getContractKitWithAgent; } });
__exportStar(require("./utils/input-validation"), exports);
__exportStar(require("./utils/key-version"), exports);
var logger_1 = require("./utils/logger");
Object.defineProperty(exports, "genSessionID", { enumerable: true, get: function () { return logger_1.genSessionID; } });
Object.defineProperty(exports, "loggerMiddleware", { enumerable: true, get: function () { return logger_1.loggerMiddleware; } });
Object.defineProperty(exports, "rootLogger", { enumerable: true, get: function () { return logger_1.rootLogger; } });
__exportStar(require("./utils/responses.utils"), exports);
