// @celo/contractkit/lib/wrappers/FeeHandler.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.FeeHandlerWrapper = exports.ExchangeProposalState = void 0;
var BaseWrapper_1 = require("./BaseWrapper");
var ExchangeProposalState;
(function (ExchangeProposalState) {
    ExchangeProposalState[ExchangeProposalState["None"] = 0] = "None";
    ExchangeProposalState[ExchangeProposalState["Proposed"] = 1] = "Proposed";
    ExchangeProposalState[ExchangeProposalState["Approved"] = 2] = "Approved";
    ExchangeProposalState[ExchangeProposalState["Executed"] = 3] = "Executed";
    ExchangeProposalState[ExchangeProposalState["Cancelled"] = 4] = "Cancelled";
})(ExchangeProposalState = exports.ExchangeProposalState || (exports.ExchangeProposalState = {}));
var FeeHandlerWrapper = /** @class */ (function (_super) {
    __extends(FeeHandlerWrapper, _super);
    function FeeHandlerWrapper() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.owner = (0, BaseWrapper_1.proxyCall)(_this.contract.methods.owner);
        _this.handleAll = (0, BaseWrapper_1.proxySend)(_this.connection, _this.contract.methods.handleAll);
        _this.burnCelo = (0, BaseWrapper_1.proxySend)(_this.connection, _this.contract.methods.burnCelo);
        return _this;
    }
    FeeHandlerWrapper.prototype.handle = function (tokenAddress) {
        return __awaiter(this, void 0, void 0, function () {
            var createExchangeProposalInner;
            return __generator(this, function (_a) {
                createExchangeProposalInner = (0, BaseWrapper_1.proxySend)(this.connection, this.contract.methods.handle);
                return [2 /*return*/, createExchangeProposalInner(tokenAddress)];
            });
        });
    };
    FeeHandlerWrapper.prototype.sell = function (tokenAddress) {
        return __awaiter(this, void 0, void 0, function () {
            var innerCall;
            return __generator(this, function (_a) {
                innerCall = (0, BaseWrapper_1.proxySend)(this.connection, this.contract.methods.sell);
                return [2 /*return*/, innerCall(tokenAddress)];
            });
        });
    };
    FeeHandlerWrapper.prototype.distribute = function (tokenAddress) {
        return __awaiter(this, void 0, void 0, function () {
            var innerCall;
            return __generator(this, function (_a) {
                innerCall = (0, BaseWrapper_1.proxySend)(this.connection, this.contract.methods.distribute);
                return [2 /*return*/, innerCall(tokenAddress)];
            });
        });
    };
    return FeeHandlerWrapper;
}(BaseWrapper_1.BaseWrapper));
exports.FeeHandlerWrapper = FeeHandlerWrapper;
