// @celo/odis-identifiers/lib/identifier.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getIdentifierHash = exports.getPrefixedIdentifier = exports.IdentifierPrefix = exports.PEPPER_SEPARATOR = void 0;
exports.PEPPER_SEPARATOR = '__';
var IdentifierPrefix;
(function (IdentifierPrefix) {
    IdentifierPrefix["NULL"] = "";
    IdentifierPrefix["PHONE_NUMBER"] = "tel";
    IdentifierPrefix["EMAIL"] = "mailto";
    IdentifierPrefix["TWITTER"] = "twit";
    IdentifierPrefix["FACEBOOK"] = "facebook";
    IdentifierPrefix["INSTAGRAM"] = "instagram";
    IdentifierPrefix["DISCORD"] = "discord";
    IdentifierPrefix["TELEGRAM"] = "telegram";
    IdentifierPrefix["SIGNAL"] = "signal";
    IdentifierPrefix["GITHUB"] = "github";
})(IdentifierPrefix || (exports.IdentifierPrefix = IdentifierPrefix = {}));
const getPrefixedIdentifier = (plaintextIdentifier, identifierPrefix) => identifierPrefix + '://' + plaintextIdentifier;
exports.getPrefixedIdentifier = getPrefixedIdentifier;
const getIdentifierHash = (sha3, plaintextIdentifier, identifierPrefix, pepper) => {
    const value = identifierPrefix === IdentifierPrefix.PHONE_NUMBER
        ? (0, exports.getPrefixedIdentifier)(plaintextIdentifier, identifierPrefix) + exports.PEPPER_SEPARATOR + pepper
        : sha3((0, exports.getPrefixedIdentifier)(plaintextIdentifier, identifierPrefix)) +
            exports.PEPPER_SEPARATOR +
            pepper;
    return sha3(value);
};
exports.getIdentifierHash = getIdentifierHash;
