// @celo/phone-number-privacy-common/lib/interfaces/endpoints.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getCombinerEndpoint = exports.getSignerEndpoint = exports.Endpoint = exports.CombinerEndpoint = exports.SignerEndpoint = exports.DomainEndpoint = exports.CombinerEndpointPNP = exports.CombinerEndpointCommon = exports.SignerEndpointPNP = exports.SignerEndpointCommon = void 0;
var SignerEndpointCommon;
(function (SignerEndpointCommon) {
    SignerEndpointCommon["METRICS"] = "/metrics";
    SignerEndpointCommon["STATUS"] = "/status";
})(SignerEndpointCommon || (exports.SignerEndpointCommon = SignerEndpointCommon = {}));
var SignerEndpointPNP;
(function (SignerEndpointPNP) {
    SignerEndpointPNP["PNP_QUOTA"] = "/quotaStatus";
    SignerEndpointPNP["PNP_SIGN"] = "/sign";
})(SignerEndpointPNP || (exports.SignerEndpointPNP = SignerEndpointPNP = {}));
var CombinerEndpointCommon;
(function (CombinerEndpointCommon) {
    CombinerEndpointCommon["METRICS"] = "/metrics";
    CombinerEndpointCommon["STATUS"] = "/status";
})(CombinerEndpointCommon || (exports.CombinerEndpointCommon = CombinerEndpointCommon = {}));
var CombinerEndpointPNP;
(function (CombinerEndpointPNP) {
    CombinerEndpointPNP["PNP_QUOTA"] = "/quotaStatus";
    CombinerEndpointPNP["PNP_SIGN"] = "/sign";
    CombinerEndpointPNP["STATUS"] = "/status";
})(CombinerEndpointPNP || (exports.CombinerEndpointPNP = CombinerEndpointPNP = {}));
var DomainEndpoint;
(function (DomainEndpoint) {
    DomainEndpoint["DOMAIN_SIGN"] = "/domain/sign";
    DomainEndpoint["DISABLE_DOMAIN"] = "/domain/disable";
    DomainEndpoint["DOMAIN_QUOTA_STATUS"] = "/domain/quotaStatus";
})(DomainEndpoint || (exports.DomainEndpoint = DomainEndpoint = {}));
exports.SignerEndpoint = Object.assign(Object.assign(Object.assign({}, SignerEndpointCommon), SignerEndpointPNP), DomainEndpoint);
exports.CombinerEndpoint = Object.assign(Object.assign(Object.assign({}, CombinerEndpointCommon), CombinerEndpointPNP), DomainEndpoint);
exports.Endpoint = Object.assign(Object.assign({}, exports.SignerEndpoint), exports.CombinerEndpoint);
function getSignerEndpoint(endpoint) {
    switch (endpoint) {
        case exports.CombinerEndpoint.DISABLE_DOMAIN:
            return exports.SignerEndpoint.DISABLE_DOMAIN;
        case exports.CombinerEndpoint.DOMAIN_QUOTA_STATUS:
            return exports.SignerEndpoint.DOMAIN_QUOTA_STATUS;
        case exports.CombinerEndpoint.DOMAIN_SIGN:
            return exports.SignerEndpoint.DOMAIN_SIGN;
        case exports.CombinerEndpoint.PNP_QUOTA:
            return exports.SignerEndpoint.PNP_QUOTA;
        case exports.CombinerEndpoint.PNP_SIGN:
            return exports.SignerEndpoint.PNP_SIGN;
        default:
            throw new Error(`No corresponding signer endpoint exists for combiner endpoint ${endpoint}`);
    }
}
exports.getSignerEndpoint = getSignerEndpoint;
function getCombinerEndpoint(endpoint) {
    switch (endpoint) {
        case exports.SignerEndpoint.DISABLE_DOMAIN:
            return exports.CombinerEndpoint.DISABLE_DOMAIN;
        case exports.SignerEndpoint.DOMAIN_QUOTA_STATUS:
            return exports.CombinerEndpoint.DOMAIN_QUOTA_STATUS;
        case exports.SignerEndpoint.DOMAIN_SIGN:
            return exports.CombinerEndpoint.DOMAIN_SIGN;
        case exports.SignerEndpoint.PNP_QUOTA:
            return exports.CombinerEndpoint.PNP_QUOTA;
        case exports.SignerEndpoint.PNP_SIGN:
            return exports.CombinerEndpoint.PNP_SIGN;
        default:
            throw new Error(`No corresponding combiner endpoint exists for signer endpoint ${endpoint}`);
    }
}
exports.getCombinerEndpoint = getCombinerEndpoint;
