// @celo/phone-number-privacy-common/lib/interfaces/responses.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.disableDomainResponseSchema = exports.domainQuotaStatusResponseSchema = exports.domainRestrictedSignatureResponseSchema = exports.PnpQuotaResponseSchema = exports.SignMessageResponseSchema = void 0;
const t = __importStar(require("io-ts"));
const PnpQuotaStatusSchema = t.type({
    performedQueryCount: t.number,
    totalQuota: t.number,
});
exports.SignMessageResponseSchema = t.union([
    t.intersection([
        t.type({
            success: t.literal(true),
            version: t.string,
            signature: t.string,
        }),
        t.partial({
            warnings: t.union([t.array(t.string), t.undefined]),
        }),
        PnpQuotaStatusSchema,
    ]),
    t.intersection([
        t.type({
            success: t.literal(false),
            version: t.string,
            error: t.string,
        }),
        t.partial({
            performedQueryCount: t.union([t.number, t.undefined]),
            totalQuota: t.union([t.number, t.undefined]),
        }),
    ]),
]);
exports.PnpQuotaResponseSchema = t.union([
    t.intersection([
        t.type({
            success: t.literal(true),
            version: t.string,
        }),
        t.partial({
            warnings: t.union([t.array(t.string), t.undefined]),
        }),
        PnpQuotaStatusSchema,
    ]),
    t.type({
        success: t.literal(false),
        version: t.string,
        error: t.string,
    }),
]);
function domainRestrictedSignatureResponseSchema(state) {
    return t.union([
        t.type({
            success: t.literal(true),
            version: t.string,
            signature: t.string,
            status: state,
        }),
        t.intersection([
            t.type({
                success: t.literal(false),
                version: t.string,
                error: t.string,
            }),
            t.partial({
                status: t.union([state, t.undefined]),
            }),
        ]),
    ]);
}
exports.domainRestrictedSignatureResponseSchema = domainRestrictedSignatureResponseSchema;
function domainQuotaStatusResponseSchema(state) {
    return t.union([
        t.type({
            success: t.literal(true),
            version: t.string,
            status: state,
        }),
        t.type({
            success: t.literal(false),
            version: t.string,
            error: t.string,
        }),
    ]);
}
exports.domainQuotaStatusResponseSchema = domainQuotaStatusResponseSchema;
function disableDomainResponseSchema(state) {
    return t.union([
        t.type({
            success: t.literal(true),
            version: t.string,
            status: state,
        }),
        t.type({
            success: t.literal(false),
            version: t.string,
            error: t.string,
        }),
    ]);
}
exports.disableDomainResponseSchema = disableDomainResponseSchema;
