// @celo/phone-number-privacy-common/lib/utils/key-version.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getResponseKeyVersion = exports.responseHasExpectedKeyVersion = exports.getRequestKeyVersion = exports.requestHasValidKeyVersion = void 0;
const __1 = require("..");
function requestHasValidKeyVersion(request, logger) {
    try {
        getRequestKeyVersion(request, logger);
        return true;
    }
    catch (err) {
        logger.debug('Error caught in requestHasValidKeyVersion');
        logger.debug(err);
        return false;
    }
}
exports.requestHasValidKeyVersion = requestHasValidKeyVersion;
function getRequestKeyVersion(request, logger) {
    const keyVersionHeader = request.headers[__1.KEY_VERSION_HEADER];
    const keyVersion = parseKeyVersionFromHeader(keyVersionHeader);
    if (keyVersion === undefined) {
        return undefined;
    }
    if (!isValidKeyVersion(keyVersion)) {
        logger.error({ keyVersionHeader, keyVersion }, __1.WarningMessage.INVALID_KEY_VERSION_REQUEST);
        throw new Error(__1.WarningMessage.INVALID_KEY_VERSION_REQUEST);
    }
    logger.info({ keyVersion }, 'Request has valid key version');
    return keyVersion;
}
exports.getRequestKeyVersion = getRequestKeyVersion;
function responseHasExpectedKeyVersion(response, expectedKeyVersion, logger) {
    let responseKeyVersion;
    try {
        responseKeyVersion = getResponseKeyVersion(response, logger);
    }
    catch (err) {
        logger.debug('Error caught in responseHasExpectedKeyVersion');
        logger.debug(err);
        return false;
    }
    if (responseKeyVersion !== expectedKeyVersion) {
        logger.error({ expectedKeyVersion, responseKeyVersion }, __1.ErrorMessage.INVALID_KEY_VERSION_RESPONSE);
        return false;
    }
    return true;
}
exports.responseHasExpectedKeyVersion = responseHasExpectedKeyVersion;
function getResponseKeyVersion(response, logger) {
    const keyVersionHeader = response.headers.get(__1.KEY_VERSION_HEADER);
    const keyVersion = parseKeyVersionFromHeader(keyVersionHeader);
    if (keyVersion === undefined) {
        return undefined;
    }
    if (!isValidKeyVersion(keyVersion)) {
        logger.error({ keyVersionHeader, keyVersion }, __1.ErrorMessage.INVALID_KEY_VERSION_RESPONSE);
        throw new Error(__1.ErrorMessage.INVALID_KEY_VERSION_RESPONSE);
    }
    logger.info({ keyVersion }, 'Response has valid key version');
    return keyVersion;
}
exports.getResponseKeyVersion = getResponseKeyVersion;
function parseKeyVersionFromHeader(keyVersionHeader) {
    if (keyVersionHeader === undefined || keyVersionHeader === null) {
        return undefined;
    }
    const keyVersionHeaderString = keyVersionHeader.toString().trim();
    if (!keyVersionHeaderString.length || keyVersionHeaderString === 'undefined') {
        return undefined;
    }
    return Number(keyVersionHeaderString);
}
function isValidKeyVersion(keyVersion) {
    return Number.isInteger(keyVersion) && keyVersion >= 0;
}
