// @celo/phone-number-privacy-common/lib/utils/logger.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.genSessionID = exports.loggerMiddleware = exports.rootLogger = void 0;
const bunyan_1 = require("bunyan");
const bunyan_debug_stream_1 = __importDefault(require("bunyan-debug-stream"));
const bunyan_gke_stackdriver_1 = require("bunyan-gke-stackdriver");
const errors_1 = require("../interfaces/errors");
const config_utils_1 = require("./config.utils");
let _rootLogger;
function rootLogger(serviceName) {
    if (_rootLogger !== undefined) {
        return _rootLogger;
    }
    const logLevel = (0, config_utils_1.fetchEnvOrDefault)('LOG_LEVEL', 'info');
    const logFormat = (0, config_utils_1.fetchEnvOrDefault)('LOG_FORMAT', 'human');
    let stream;
    switch (logFormat) {
        case 'stackdriver':
            stream = (0, bunyan_gke_stackdriver_1.createStream)(bunyan_1.levelFromName[logLevel]);
            break;
        case 'json':
            stream = { stream: process.stdout, level: logLevel };
            break;
        default:
            stream = { level: logLevel, stream: (0, bunyan_debug_stream_1.default)() };
            break;
    }
    _rootLogger = (0, bunyan_1.createLogger)({
        name: serviceName !== null && serviceName !== void 0 ? serviceName : '',
        serializers: bunyan_1.stdSerializers,
        streams: [stream],
    });
    return _rootLogger;
}
exports.rootLogger = rootLogger;
function loggerMiddleware(serviceName) {
    return (req, res, next) => {
        const requestLogger = rootLogger(serviceName).child({
            endpoint: req.path,
            sessionID: req.body.sessionID, // May be undefined
        });
        res.locals.logger = requestLogger;
        if (!req.body.sessionID && req.path !== '/metrics' && req.path !== '/status') {
            requestLogger.info(errors_1.WarningMessage.MISSING_SESSION_ID);
        }
        if (next) {
            next();
        }
        return requestLogger;
    };
}
exports.loggerMiddleware = loggerMiddleware;
function genSessionID() {
    return Math.random().toString(36).substring(2, 15) + Math.random().toString(36).substring(2, 15);
}
exports.genSessionID = genSessionID;
